<a href="https://colab.research.google.com/github/sreelakshmisajeevan-854/Data-Science-Assignments/blob/main/Data_Preprocessing_and_Feature_Engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# 1.Data Exploration and Preprocessing:

In [ ]:
df=pd.read_csv('/content/drive/MyDrive/Datasets/adult_with_headers (1).csv')

In [ ]:
df.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [ ]:
df.shape

(32561, 15)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [ ]:
df.describe()

,age,fnlwgt,education_num,capital_gain,capital_loss,hours_per_week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


Inference


*   The dataset contains 32,561 observations and 15 features.
*   The dataset contains both numerical (int64) and categorical (object) variables.

*   The average age is 38.58 years, with most individuals between 28 and 48 years.
*   fnlwgt shows high variability, indicating possible outliers.

*   Most individuals have an education level between 9 and 12.
*   Most values are 0 for capital_gain and	capital_loss, indicating a highly right-skewed distribution with a few large values.

*   Most individuals work around 40 hours per week, with a few extreme values (up to 99 hours).










In [ ]:
#Find all unique values
df.isnull().sum()

,0
age,0
workclass,0
fnlwgt,0
education,0
education_num,0
marital_status,0
occupation,0
relationship,0
race,0
sex,0


In [ ]:
#Find all unique values
df['workclass'].unique()

array([' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
       ' Local-gov', ' ?', ' Self-emp-inc', ' Without-pay',
       ' Never-worked'], dtype=object)

In [ ]:

for col in df.select_dtypes(include='object').columns:
    print(col)
    print(df[col].unique())
    print("-" * 30)

workclass
[' State-gov' ' Self-emp-not-inc' ' Private' ' Federal-gov' ' Local-gov'
 ' ?' ' Self-emp-inc' ' Without-pay' ' Never-worked']
------------------------------
education
[' Bachelors' ' HS-grad' ' 11th' ' Masters' ' 9th' ' Some-college'
 ' Assoc-acdm' ' Assoc-voc' ' 7th-8th' ' Doctorate' ' Prof-school'
 ' 5th-6th' ' 10th' ' 1st-4th' ' Preschool' ' 12th']
------------------------------
marital_status
[' Never-married' ' Married-civ-spouse' ' Divorced'
 ' Married-spouse-absent' ' Separated' ' Married-AF-spouse' ' Widowed']
------------------------------
occupation
[' Adm-clerical' ' Exec-managerial' ' Handlers-cleaners' ' Prof-specialty'
 ' Other-service' ' Sales' ' Craft-repair' ' Transport-moving'
 ' Farming-fishing' ' Machine-op-inspct' ' Tech-support' ' ?'
 ' Protective-serv' ' Armed-Forces' ' Priv-house-serv']
------------------------------
relationship
[' Not-in-family' ' Husband' ' Wife' ' Own-child' ' Unmarried'
 ' Other-relative']
------------------------------
race
[' W

In [ ]:
#Check missing values
(df == ' ?').sum()

,0
age,0
workclass,1836
fnlwgt,0
education,0
education_num,0
marital_status,0
occupation,1843
relationship,0
race,0
sex,0


There are a large number of missing values for workclass, occupation and native_country.

In [ ]:
#Replace ' ?' with NaN
df = df.replace(' ?', np.nan)

In [ ]:
df.isnull().sum()

,0
age,0
workclass,1836
fnlwgt,0
education,0
education_num,0
marital_status,0
occupation,1843
relationship,0
race,0
sex,0


In [ ]:
#Fill the missing values
for col in ['workclass', 'occupation', 'native_country']:
    df[col].fillna(df[col].mode()[0], inplace=True)

In [ ]:
df.isnull().sum()

,0
age,0
workclass,0
fnlwgt,0
education,0
education_num,0
marital_status,0
occupation,0
relationship,0
race,0
sex,0


Scaling

In [ ]:
#Standard scaling
df.select_dtypes(include=['int64', 'float64']).columns

Index(['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss',
       'hours_per_week'],
      dtype='object')

In [ ]:
num_cols = ['age', 'fnlwgt', 'education_num',
            'capital_gain', 'capital_loss',
            'hours_per_week']

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df_standard = df.copy()
df_standard[num_cols] = scaler.fit_transform(df_standard[num_cols])
df_standard.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,0.030671,State-gov,-1.063611,Bachelors,1.134739,Never-married,Adm-clerical,Not-in-family,White,Male,0.148453,-0.21666,-0.035429,United-States,<=50K
1,0.837109,Self-emp-not-inc,-1.008707,Bachelors,1.134739,Married-civ-spouse,Exec-managerial,Husband,White,Male,-0.145920,-0.21666,-2.222153,United-States,<=50K
2,-0.042642,Private,0.245079,HS-grad,-0.420060,Divorced,Handlers-cleaners,Not-in-family,White,Male,-0.145920,-0.21666,-0.035429,United-States,<=50K
3,1.057047,Private,0.425801,11th,-1.197459,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,-0.145920,-0.21666,-0.035429,United-States,<=50K
4,-0.775768,Private,1.408176,Bachelors,1.134739,Married-civ-spouse,Prof-specialty,Wife,Black,Female,-0.145920,-0.21666,-0.035429,Cuba,<=50K


In [ ]:
#Min-Max Scaling
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df_minmax = df.copy()
df_minmax[num_cols] = scaler.fit_transform(df_minmax[num_cols])
df_minmax.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,0.301370,State-gov,0.044302,Bachelors,0.800000,Never-married,Adm-clerical,Not-in-family,White,Male,0.02174,0.0,0.397959,United-States,<=50K
1,0.452055,Self-emp-not-inc,0.048238,Bachelors,0.800000,Married-civ-spouse,Exec-managerial,Husband,White,Male,0.00000,0.0,0.122449,United-States,<=50K
2,0.287671,Private,0.138113,HS-grad,0.533333,Divorced,Handlers-cleaners,Not-in-family,White,Male,0.00000,0.0,0.397959,United-States,<=50K
3,0.493151,Private,0.151068,11th,0.400000,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0.00000,0.0,0.397959,United-States,<=50K
4,0.150685,Private,0.221488,Bachelors,0.800000,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0.00000,0.0,0.397959,Cuba,<=50K


Standard Scaling


*   Standard Scaling is preferred for algorithms that depend on the distribution and variance of the data
*   Standard Scaling is preferred when the numerical features have different scales and the machine learning algorithm assumes that the data is approximately normally distributed.
*   It transforms the data to have a mean of 0 and a standard deviation of 1.

Min-Max Scaling


*   Min-Max Scaling is preferred when the data needs to be transformed into a fixed range, typically 0 to 1.
*   It preserves the relationships between the original values while ensuring that all features have the same range.
*   Min-Max Scaling is more sensitive to outliers because the minimum and maximum values determine the scaling range.







# 2. Encoding Techniques:

In [ ]:
for col in df.select_dtypes(include='object').columns:
    print(col, ":", df[col].nunique())

workclass : 8
education : 16
marital_status : 7
occupation : 14
relationship : 6
race : 5
sex : 2
native_country : 41
income : 2


One-Hot Encoding

In [ ]:
#One-Hot Encoding only to columns with less than 5 categories
from sklearn.preprocessing import OneHotEncoder
dfOhe=df.copy()
ohe=OneHotEncoder(sparse_output=False,drop='first')
dfEncode1=ohe.fit_transform(dfOhe[['sex','income']])
dfEncode1=pd.DataFrame(dfEncode1,columns=ohe.get_feature_names_out())
dfEncode1.head()

,sex_ Male,income_ >50K
0,1.0,0.0
1,1.0,0.0
2,1.0,0.0
3,1.0,0.0
4,0.0,0.0


Label Encoding

In [ ]:
#Label Encoding for variables with more than 5 categories
cols = ['workclass',
        'education',
        'marital_status',
        'occupation',
        'relationship',
        'native_country']

In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df_label = df.copy()
for col in cols:
    df_label[col] = le.fit_transform(df_label[col])
df_label.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,39,6,77516,9,13,4,0,1,White,Male,2174,0,40,38,<=50K
1,50,5,83311,9,13,2,3,0,White,Male,0,0,13,38,<=50K
2,38,3,215646,11,9,0,5,1,White,Male,0,0,40,38,<=50K
3,53,3,234721,1,7,2,5,0,Black,Male,0,0,40,38,<=50K
4,28,3,338409,9,13,2,9,5,Black,Female,0,0,40,4,<=50K


Pros and Cons of One-Hot Encoding


```
Pros:
*   Does not introduce any order among categories.
*   Suitable for nominal categorical variables.
*   Improves performance of many machine learning algorithms.

Cons:
*   Creates many new columns when there are many categories.
*   Increases memory usage.
*   Can make the dataset sparse and slower to process.

```
Pros and Cons of Label Encoding


```
Pros:


*   Simple and easy to implement.
*   Uses less memory since only one column is created.
*   Suitable for ordinal variables and tree-based models.


Cons:


*   Introduces an artificial order among categories.
*   May negatively affect algorithms like Logistic Regression, KNN, and SVM if used on nominal variables.

```





# 3. Feature Engineering:

Create  2 new features

In [ ]:
df['total_capital'] = df['capital_gain'] - df['capital_loss']
df['total_capital'].head()

,total_capital
0,2174
1,0
2,0
3,0
4,0


Rationale

*   Combines capital gain and capital loss into a single measure.
*   Represents an individual's overall capital income.



In [ ]:
df['work_intensity'] = df['hours_per_week'] / 40
df['work_intensity'].head()

,work_intensity
0,1.000
1,0.325
2,1.000
3,1.000
4,1.000


Rationale


*   A standard full-time work week is approximately 40 hours.
*   This feature may help capture the relationship between working hours and income.



Apply Log Transformation for capital_gain


In [ ]:
df['capital_gain_log'] = np.log1p(df['capital_gain'])

In [ ]:
df['capital_gain_log'].head()

,capital_gain_log
0,7.684784
1,0.000000
2,0.000000
3,0.000000
4,0.000000


The capital_gain feature is highly right-skewed because most values are 0 while a few are extremely large.Log transformation reduces the effect of extremely large values.It makes the distribution more balanced.